### PR TITLE
[Plugin] Enhanced custom validation

### PR DESCRIPTION
**Pull request by SchrodingersGat**
This PR provides a significant improvement to the `ValidationMixin` plugin - allowing greater flexibility in custom validation of database models

- Adds a `validate_model_instance` method to the `ValidationMixin` plugin
- Exposes "deltas" to the method - allowing plugins to validate based on which fields have been changed
- Exposes *most* database models to the this method using the `PluginValidationMixin` model mixin class
- Updates docs and adds more clear examples

### CODE REVIEW

File: InvenTree/plugin/helpers.py [hunk 1]
```diff
@@ -220,6 +220,37 @@ def get_plugins(pkg, baseclass, path=None):
...
+
+    if name and cfg.name != name:
+        # Update the name if it has changed
+        try:
+            cfg.name = name
+            cfg.save()
+        except Exception as e:
+            logger.exception(f'Failed to update plugin name: {e}')
+
+    return cfg
```
Comment (matmair) [inline InvenTree/plugin/helpers.py:251]: Why is this moved out of the registry? IMO it makes more sense there as the registry is always made available under a constant import path

File: InvenTree/stock/models.py [hunk 2]
```diff
@@ -26,20 +26,13 @@
...
 from InvenTree.fields import InvenTreeModelMoneyField, InvenTreeURLField
-from InvenTree.models import (
-    InvenTreeAttachment,
-    InvenTreeBarcodeMixin,
-    InvenTreeNotesMixin,
-    InvenTreeTree,
-    MetadataMixin,
-    extract_int,
-)
 from InvenTree.status_codes import (
```
Comment (matmair) [inline InvenTree/stock/models.py:36]: Question: Why is this import changed?

File: InvenTree/part/models.py [hunk 3]
```diff
@@ -37,6 +37,7 @@
 import common.settings
 import InvenTree.conversion
 import InvenTree.fields
+import InvenTree.models
```
Comment (matmair) [inline InvenTree/part/models.py:40]: Question: Why is this import changed?

File: InvenTree/plugin/helpers.py (hunk 1 above)
Comment (SchrodingersGat) [reply InvenTree/plugin/helpers.py]: Fair enough!
